### PR TITLE
Fix vipllava for generation

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

When working on [this PR](https://github.com/huggingface/transformers/pull/29853), it was found that VipLlava fails when generating with kv cache, because of incorrectly indexing `past_kv_length`. This PR fixes it by indexing the past length as `-2`. Other Llava models work correctly.

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Thanks! Do you know if our slow CI caught this? Otherwise let's add a test!

**Comment by zucchini-nlp** [comment]
Yes, the `test_small_model_integration_test` should have caught it. Also, soon we will have all GenerationTesterMixin tests running for vision language models, to catch this type of errors without slow CI.

**Review by ArthurZucker — COMMENTED** [review]
I am a bit surprised by the other model working as expected because other llava based model do use [-1] which is the `head_dim` no?

**Comment by zucchini-nlp** [comment]
Yeah, that's because other models index only the first head dim in [this line ](https://github.com/huggingface/transformers/blob/ae49b218c3d718df90d8e4a109016450fb8f0632/src/transformers/models/llava/modeling_llava.py#L442), while VipLlava indexes the first head.

I found why we need this head indexing hack and tried generating with padded inputs, I did not see any difference between indexing first head or first head dimension.

File: src/transformers/models/vipllava/modeling_vipllava.py [hunk 1]
```diff
@@ -447,7 +447,7 @@ def forward(
                     batch_index, non_attended_tokens = torch.where(first_layer_past_key_value.float().sum(-1) == 0)
```
Comment (ArthurZucker) [inline src/transformers/models/vipllava/modeling_vipllava.py:447]: I am guessing we could also just do: 
```python 
                first_layer_past_key_value = past_key_values[0][0][:, :, :, 0]

                # Sum all dimensions of head_dim (-2) to avoid random errors such as: https://github.com/huggingface/transformers/pull/28032#issuecomment-1863691941
                batch_index, non_attended_tokens = torch.where(first_layer_past_key_value.float().sum(-2) == 0)
```
no?

Comment (ArthurZucker) [reply src/transformers/models/vipllava/modeling_vipllava.py:447]: just want to be consistent!

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/models/vipllava/modeling_vipllava.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/vipllava/modeling_vipllava.py:447]: sure, let me change it for consistency. As far as I understand, the hack is to not rely on only one value to be zero, which caused random errors.

**Comment by zucchini-nlp** [comment]
Made Llava model code consistent and ran all tests (+slow). Some that were failing do not have anything to do with the current changes, I confirmed they were failing long time before

File: tests/models/llava/test_modeling_llava.py [hunk 2]
```diff
@@ -403,7 +403,7 @@ def test_small_model_integration_test_llama(self):
...
+        EXPECTED_DECODED_TEXT = "USER:  \nWhat are the things I should be cautious about when I visit this place?\nASSISTANT: When visiting this place, which appears to be a dock or pier extending over a body of water, there are a few things to be cautious about. First, be aware of the surroundings and potential hazards, such as slippery surfaces, uneven ground, or any obstacles in the water. Second, be mindful of the weather conditions, as sudden changes in weather can make the dock or pier unsafe to use. Third, be cautious of the water depth and any underwater hazards, such as rocks or debris, that could pose a risk to your safety. Lastly, be respectful of the environment and other visitors, and follow any rules or guidelines posted at the dock or pier."  # fmt: skip
```
Comment (ArthurZucker) [inline tests/models/llava/test_modeling_llava.py:406]: Which hardware did you run this on ? should not be failing! but maybe it T4 vs A100

File: src/transformers/models/llava_next/modeling_llava_next.py
Comment (ArthurZucker) [inline src/transformers/models/llava_next/modeling_llava_next.py:1]: nice catch

**Review by ArthurZucker — COMMENTED** [review]
Thanks

File: src/transformers/models/vipllava/modeling_vipllava.py [hunk 3]
```diff
@@ -441,10 +441,10 @@ def forward(
                 if past_key_values is not None and pixel_values is not None and input_ids.shape[1] == 1:
                     # Retrieve the first layer to inspect the logits and mask out the hidden states
                     # that are set to 0
-                    first_layer_past_key_value = past_key_values[0][0][:, 0, :, :]
+                    first_layer_past_key_value = past_key_values[0][0][:, :, :, 0]
```
Comment (gante) [inline src/transformers/models/vipllava/modeling_vipllava.py:444]: This seems like an opportunity for `# Copied from` (not to be fixed in this PR, but in the future)

File: tests/models/llava/test_modeling_llava.py (hunk 2 above)
Comment (gante) [reply tests/models/llava/test_modeling_llava.py:406]: Related question: if `llava` was not touched in this PR and our daily slow CI is not complaining, why was this test changed? 🤔

Comment (zucchini-nlp) [reply tests/models/llava/test_modeling_llava.py:406]: Hmm, I was running the test on A100. If daily CI is passing, then prob it does not need change. Anyway, it's interesting for me that some tests are reliant on the hardware, which means a contributor might change failing tests without knowing that it's correct. 

Should I just revert changes then?

File: src/transformers/models/vipllava/modeling_vipllava.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/models/vipllava/modeling_vipllava.py:444]: Noted for the next PR :)

**Review by gante — APPROVED** [review]
LGTM 👍 

Let's revert the changes in `tests/models/llava/test_modeling_llava.py`, then we can merge :)
